### PR TITLE
Add parameters to binding/instance creation

### PR DESCRIPTION
**Pull request by vaikas-google**
Follow onto the #366 to add parameter passing into the WIP controller.

### CODE REVIEW

File: pkg/controller/wip/controller.go [hunk 1]
```diff
@@ -382,11 +383,26 @@ func (c *controller) reconcileInstance(instance *v1alpha1.Instance) {
...
+	parameters := make(map[string]interface{})
+	if len(instance.Spec.Parameters.Raw) > 0 {
+		err = yaml.Unmarshal([]byte(instance.Spec.Parameters.Raw), &parameters)
+		if err != nil {
+			glog.Errorf("Failed to unmarshal Instance parameters\n%s\n %v", instance.Spec.Parameters, err)
+			c.updateInstanceCondition(
+				instance,
+				v1alpha1.InstanceConditionReady,
+				v1alpha1.ConditionFalse,
+				"ErrorWithParameters",
```
Comment (arschles) [inline pkg/controller/wip/controller.go]: can you put this into a `const`?

File: pkg/controller/wip/controller.go [hunk 2]
```diff
@@ -552,11 +568,26 @@ func (c *controller) reconcileBinding(binding *v1alpha1.Binding) {
 	glog.V(4).Infof("Creating client for Broker %v, URL: %v", broker.Name, broker.Spec.URL)
 	brokerClient := c.brokerClientCreateFunc(broker.Name, broker.Spec.URL, username, password)
 
-	// TODO: parameters
+	parameters := make(map[string]interface{})
+	if len(binding.Spec.Parameters.Raw) > 0 {
+		err = yaml.Unmarshal([]byte(binding.Spec.Parameters.Raw), &parameters)
+		if err != nil {
```
Comment (arschles) [inline pkg/controller/wip/controller.go]: can you break the above 4 lines into a reusable function?

also, you can collapse the above 2 lines into:

```go
if err := yaml.Unmarshal(...); err != nil {
```

File: pkg/controller/wip/controller.go (hunk 1 above)
Comment (vaikas-google) [inline pkg/controller/wip/controller.go]: Yeah, I can do that, just looking at the file it seemed to not use consts so was trying to be consistent with the file convention.

Comment (arschles) [inline pkg/controller/wip/controller.go]: go forth and `const`!

File: pkg/controller/wip/controller.go (hunk 2 above)
Comment (vaikas-google) [inline pkg/controller/wip/controller.go]: done.

**Comment by vaikas-google** [comment]
Ok, added unit tests, there's more that need to be added, but I'd like to do that in a follow on PR. In any case, PTAL.

File: pkg/controller/wip/controller.go [hunk 3]
```diff
@@ -552,13 +571,26 @@ func (c *controller) reconcileBinding(binding *v1alpha1.Binding) {
...
 
 	request := &brokerapi.BindingRequest{
-		ServiceID: serviceClass.OSBGUID,
-		PlanID:    servicePlan.OSBGUID,
+		ServiceID:  serviceClass.OSBGUID,
+		PlanID:     servicePlan.OSBGUID,
+		Parameters: parameters,
 	}
 
+	// TODO(vaikas): Shouldn't the binding.Spec.OSBGUID get generated by the controller?
```
Comment (pmorie) [inline pkg/controller/wip/controller.go]: We have a default that generates it.  It should already be set here.

File: pkg/controller/wip/controller.go [hunk 4]
```diff
@@ -718,3 +750,13 @@ func convertServicePlans(plans []brokerapi.ServicePlan) []v1alpha1.ServicePlan {
 	}
 	return ret
 }
+
+func unmarshalParameters(in []byte) (map[string]interface{}, error) {
```
Comment (pmorie) [inline pkg/controller/wip/controller.go]: So, we get a raw yaml blob in the `Instance` and `Binding` resources, unmarshal into `map[string]interface{}` for our go OSB library, which in turn serializes them into json in the request we send to the broker, accurate?

File: pkg/controller/wip/controller_test.go [hunk 5]
```diff
@@ -31,34 +32,57 @@ import (
 	"k8s.io/kubernetes/pkg/client/testing/core"
 )
 
-func TestReconcileBroker(t *testing.T) {
-	fakeKubeClient, fakeCatalogClient, fakeBrokerCatalog, _, _, testController, _, stopCh := newTestController(t)
+// broker used in most of the tests that need a broker
+var broker = &v1alpha1.Broker{
```
Comment (pmorie) [inline pkg/controller/wip/controller_test.go]: I agree with this refactor, but prefer these as functions that return objects so that a test can't accidentally mutate one and clobber the other tests.

File: pkg/controller/wip/controller_test.go [hunk 6]
```diff
@@ -185,6 +260,248 @@ func TestReconcileInstanceHappyLand(t *testing.T) {
...
+
+	if e, a := v1alpha1.InstanceConditionReady, updateObject.Status.Conditions[0].Type; e != a {
+		t.Fatalf("Unexpected condition type: expected %v, got %v", e, a)
+	}
+
+	if e, a := v1alpha1.ConditionTrue, updateObject.Status.Conditions[0].Status; e != a {
+		t.Fatalf("Unexpected condition status: expected %v, got %v", e, a)
+	}
+
+	// TODO(vaikas): Since the Instance OSBGUID is not set, it gets marked as "" which is certainly
```
Comment (pmorie) [inline pkg/controller/wip/controller_test.go]: We need to just hard-code a GUID for `ServiceClass`, `ServicePlan`, `Instance`, and `Binding` in these tests.

File: pkg/controller/wip/controller_test.go [hunk 7]
```diff
@@ -185,6 +260,248 @@ func TestReconcileInstanceHappyLand(t *testing.T) {
...
+	// not right, but it's a bit unclear if this should get assigned by a controller or
+	// if the test instance above should set it. I'll tackle this in a follow on CL since this
+	// is not directly related to my change re: parameters.
+	if si, ok := fakeInstanceClient.Instances[""]; !ok {
+		t.Fatalf("Did not find the created Instance in fakeInstanceClient after creation")
+	} else {
+		if len(si.Parameters) > 0 {
+			t.Fatalf("Unexpected parameters, expected none, got %+v", si.Parameters)
```
Comment (pmorie) [inline pkg/controller/wip/controller_test.go]: It would be good to verify that we got the params that we expected here.

File: pkg/controller/wip/controller.go (hunk 4 above)
Comment (vaikas-google) [inline pkg/controller/wip/controller.go]: yes.

File: pkg/controller/wip/controller_test.go [hunk 8]
```diff
@@ -31,34 +32,57 @@ import (
...
+	Args map[string]string `json:"args"`
+}
+
+type bindingParameters struct {
+	Name string   `json:"name"`
+	Args []string `json:"args"`
+}
+
+func TestReconcileBroker(t *testing.T) {
+	fakeKubeClient, fakeCatalogClient, fakeBrokerCatalog, _, _, testController, _, stopCh := newTestController(t)
```
Comment (arschles) [inline pkg/controller/wip/controller_test.go]: can you call `defer close(stopCh)`, instead of at the end of the test. same with the other tests

not blocking for this PR, however, since it's in many other tests. If you decide not to do it, can you create an issue for it?

File: pkg/controller/wip/controller_test.go (hunk 6 above)
Comment (vaikas-google) [inline pkg/controller/wip/controller_test.go]: Ok, so who's job is it to actually fill in the GUID, etc.? Isn't that the job of the controller?

File: pkg/controller/wip/controller_test.go (hunk 7 above)
Comment (vaikas-google) [inline pkg/controller/wip/controller_test.go]: ?? We don't expect parameters here. It's elsewhere, or do I misunderstand what you mean?

File: pkg/controller/wip/controller_test.go (hunk 6 above)
Comment (pmorie) [inline pkg/controller/wip/controller_test.go]: It's the API server's job to create GUIDs for: `Instance`, `Binding`.  GUIDs for `ServiceClass` and `ServicePlan` come from the broker.  So, in this test, we need to just hard-code some GUIDs in the objects we're using in the test.

File: pkg/controller/wip/controller_test.go (hunk 5 above)
Comment (vaikas-google) [inline pkg/controller/wip/controller_test.go]: done

File: pkg/controller/wip/controller.go [hunk 9]
```diff
@@ -382,16 +387,30 @@ func (c *controller) reconcileInstance(instance *v1alpha1.Instance) {
...
+		ServiceID:  serviceClass.OSBGUID,
+		PlanID:     servicePlan.OSBGUID,
+		Parameters: parameters,
 	}
 
 	// TODO: handle async provisioning
 
 	glog.V(4).Infof("Provisioning a new Instance %v/%v of ServiceClass %v at Broker %v", instance.Namespace, instance.Name, serviceClass.Name, broker.Name)
+	// TODO(vaikas): Who populates the OSBGUID here? I thought controller creates a GUID, but I don't
+	// see it being generated anywhere...
```
Comment (pmorie) [inline pkg/controller/wip/controller.go]: We already have a default that makes GUIDs for `Instance` and `Binding`.  The issue is that the objects we're creating in code in this test don't go through the defaults since they don't go through the API server.  So, we need to fill the GUID fields in ourselves for this test and others like it.

File: pkg/controller/wip/controller_test.go (hunk 7 above)
Comment (pmorie) [inline pkg/controller/wip/controller_test.go]: Oh, I'm sorry, I was thinking this was the test case where we do pass parameters.  My bad!

File: pkg/controller/wip/controller_test.go [hunk 10]
```diff
@@ -185,6 +260,248 @@ func TestReconcileInstanceHappyLand(t *testing.T) {
...
+
+	// Verify parameters are what we'd expect them to be, basically name, array with two values in it.
+	if len(updateObject.Spec.Parameters.Raw) == 0 {
+		t.Fatalf("Parameters was unexpectedly empty")
+	}
+	// TODO(vaikas): Implement the storing logic in the fake Binding Client so that it stores
+	// something meaningful there. For now, it just stores a struct making the validation a bit
+	// wonky.
+	if _, ok := fakeBindingClient.Bindings[":"]; !ok {
+		t.Fatalf("Did not find the created Binding in fakeInstanceBinding after creation")
```
Comment (pmorie) [inline pkg/controller/wip/controller_test.go]: We should verify that we got the parameters that we expected here and in the test case where we provision with parameters.

Comment (vaikas-google) [inline pkg/controller/wip/controller_test.go]: Yes, we should :) See the comment why not and why I'd rather do that in a follow on PR. ok?

Comment (pmorie) [inline pkg/controller/wip/controller_test.go]: What, you don't want to block this PR on learning go reflection? :)

I'm cool with a follow-up.

Comment (vaikas-google) [inline pkg/controller/wip/controller_test.go]: Hm, perhaps I misunderstood what this does then:
https://github.com/kubernetes-incubator/service-catalog/blob/master/pkg/brokerapi/fake/fake.go#L182
Seems like it's not storing anything useful there, but ??

File: pkg/controller/wip/controller.go [hunk 11]
```diff
@@ -382,16 +387,30 @@ func (c *controller) reconcileInstance(instance *v1alpha1.Instance) {
 	glog.V(4).Infof("Creating client for Broker %v, URL: %v", broker.Name, broker.Spec.URL)
 	brokerClient := c.brokerClientCreateFunc(broker.Name, broker.Spec.URL, username, password)
 
-	// TODO: parameters
+	parameters, err := unmarshalParameters(instance.Spec.Parameters.Raw)
```
Comment (vaikas-google) [inline pkg/controller/wip/controller.go]: well, since I was asked to pull it into a method, I moved it there. Otherwise, there's really no point abstracting away a call to yaml.unmarshal. Also, since the method returns a blank parameters, don't need to initialize here, ok?

File: pkg/controller/wip/controller_test.go (hunk 8 above)
Comment (vaikas-google) [inline pkg/controller/wip/controller_test.go]: done. and added a bug to track others too...

**Comment by vaikas-google** [comment]
Thanks for the review, I believe I've addressed all the comments (mea culpa if I missed any), PTAL.

**Comment by pmorie** [comment]
Checking to see if we can bounce travis

**Comment by MHBauer** [comment]
Bounce?

File: pkg/controller/wip/controller.go [hunk 12]
```diff
@@ -342,7 +347,7 @@ func (c *controller) reconcileInstance(instance *v1alpha1.Instance) {
 
 	servicePlan := findServicePlan(instance.Spec.PlanName, serviceClass.Plans)
 	if servicePlan == nil {
-		glog.Errorf("Instance %v/%v references a non-existent ServicePlan %v on ServiceClass %v", instance.Namespace, instance.Name, servicePlan.Name, serviceClass.Name)
+		glog.Errorf("Instance %v/%v references a non-existent ServicePlan %v on ServiceClass %v", instance.Namespace, instance.Name, instance.Spec.PlanName, serviceClass.Name)
```
Comment (duglin) [inline pkg/controller/wip/controller.go]: can you either wrap the %v's with quotes ('xx') or change some to %q so golang does it for you. Names can sometimes be words that make the sentence hard to parse for newbies because they blend into a normal english sentence.  Wrapping them with quotes will remove that risk.

File: pkg/controller/wip/controller.go (hunk 11 above)
Comment (duglin) [inline pkg/controller/wip/controller.go]: hmmm, if we have no params I think unmarshalParameters() will still generate a map, but an empty one. Which I think will result in an empty Parameters property, not a nil one. Will this cause us to generate an empty json field instead of skipping the field all together in the json?

File: pkg/controller/wip/controller.go (hunk 12 above)
Comment (duglin) [inline pkg/controller/wip/controller.go]: this applies to all use of %v below too.

File: pkg/controller/wip/controller.go [hunk 13]
```diff
@@ -382,11 +387,23 @@ func (c *controller) reconcileInstance(instance *v1alpha1.Instance) {
...
-	// TODO: parameters
+	parameters, err := unmarshalParameters(instance.Spec.Parameters.Raw)
+	if err != nil {
+		glog.Errorf("Failed to unmarshal Instance parameters\n%s\n %v", instance.Spec.Parameters, err)
+		c.updateInstanceCondition(
+			instance,
+			v1alpha1.InstanceConditionReady,
+			v1alpha1.ConditionFalse,
+			errorWithParameters,
+			"Error unmarshaling instance parameters",
```
Comment (duglin) [inline pkg/controller/wip/controller.go]: Can we add the error text into the condition so that people know what happened? Or are conditions not supposed to be that verbose?

File: pkg/controller/wip/controller.go [hunk 14]
```diff
@@ -486,12 +503,12 @@ func (c *controller) reconcileBinding(binding *v1alpha1.Binding) {
...
 	if err != nil {
-		glog.Errorf("Binding %v/%v references a non-existent Instance %v/%v", binding.Namespace, binding.Name, binding.Namespace, instance.Name)
+		glog.Errorf("Binding %v/%v references a non-existent Instance %v/%v", binding.Namespace, binding.Name, binding.Namespace, binding.Spec.InstanceRef.Name)
 		c.updateBindingCondition(
 			binding,
 			v1alpha1.BindingConditionReady,
 			v1alpha1.ConditionFalse,
-			"ReferencesNonexistentServiceClass",
+			"ReferencesNonexistentInstance",
 			"The binding references an Instance that does not exist",
```
Comment (duglin) [inline pkg/controller/wip/controller.go]: Adding the name of the unfound instance to the condition would be nice.

File: pkg/controller/wip/controller_test.go (hunk 8 above)
Comment (duglin) [inline pkg/controller/wip/controller_test.go]: destructors should would be nice :-)

File: pkg/controller/wip/controller_test.go [hunk 15]
```diff
@@ -96,59 +134,113 @@ func TestReconcileBroker(t *testing.T) {
...
+
+	actions := filterActions(fakeCatalogClient.Actions())
+	if e, a := 1, len(actions); e != a {
+		t.Logf("%+v\n", actions)
+		t.Fatalf("Unexpected number of actions: expected %v, got %v", e, a)
+	}
+
+	// There should only be one action that says it failed because no such class exists.
+	updateAction := actions[0].(core.UpdateAction)
+	if e, a := "update", updateAction.GetVerb(); e != a {
```
Comment (duglin) [inline pkg/controller/wip/controller_test.go]: This is an interesting way to write an if-statement. I suspect you did it so that you can reuse 'e' and 'a' in the t.Fatalf() but it may make it harder for newbies to follow.  Won't ask for a change since it might just be me, but did want to point it out for others to comment if they're worried too.

Honestly, I can't decide if its "cool" or "too cool that its hard to follow"  :-)

File: pkg/controller/wip/controller_test.go [hunk 16]
```diff
@@ -96,59 +134,113 @@ func TestReconcileBroker(t *testing.T) {
...
-	broker := &v1alpha1.Broker{
-		ObjectMeta: v1.ObjectMeta{Name: "test-broker"},
-		Spec: v1alpha1.BrokerSpec{
-			URL:     "https://example.com",
-			OSBGUID: "OSBGUID field",
+	testController.reconcileInstance(instance)
+
+	actions := filterActions(fakeCatalogClient.Actions())
+	if e, a := 1, len(actions); e != a {
+		t.Logf("%+v\n", actions)
```
Comment (duglin) [inline pkg/controller/wip/controller_test.go]: I don't think we need the log here. it would be better if this info were part of the Fatal msg so that it shows up on the testing output.

File: pkg/controller/wip/controller.go (hunk 13 above)
Comment (pmorie) [inline pkg/controller/wip/controller.go]: It's fine and good to have messages that are verbose, but the rest of the messages also lack this information.  I would rather this go in now and someone take a pass through all the messages on conditions later.

File: pkg/controller/wip/controller.go (hunk 12 above)
Comment (pmorie) [inline pkg/controller/wip/controller.go]: That's a fair point, but not related strictly to this PR.  One other thing that will happen, and has already happened in our codebase, is that everyone who adds a log statement that prints information about an object will do it differently.  When I was still working in the kubelet day to day, we had discussed having methods that gave you the printed representation for logs, and always using those so that the format was consistent everywhere and could be changed everywhere easily.

The only thing @vaikas-google is changing here is that there's a nil passed (the servicePlan, which couldn't be found) instead of meaningful information (the service plan named in the object's spec).  

I've created https://github.com/kubernetes-incubator/service-catalog/issues/451 for this

File: pkg/controller/wip/controller.go (hunk 13 above)
Comment (pmorie) [inline pkg/controller/wip/controller.go]: Created https://github.com/kubernetes-incubator/service-catalog/issues/452

File: pkg/controller/wip/controller_test.go (hunk 15 above)
Comment (pmorie) [inline pkg/controller/wip/controller_test.go]: It's not about being cool, it's about making repetitive if-statements in tests that are compact and easy to read.  It's also idiomatic go, and builds character.

File: pkg/controller/wip/controller_test.go (hunk 16 above)
Comment (pmorie) [inline pkg/controller/wip/controller_test.go]: I think much cleanup of this type is needed in tests, and I would like to do this after we finish MVP 1 objectives.

File: pkg/controller/wip/controller.go (hunk 11 above)
Comment (pmorie) [inline pkg/controller/wip/controller.go]: I realized last night while doing some cleanups to the API that the parameters field should really be a pointer so that we can easily test for 'unset'.  I'll clean this up in a followup today after this is merged.

**Comment by pmorie** [comment]
Thanks for the patch @vaikas-google

**Comment by duglin** [comment]
I guess there's no point in continuing my review....
